Common/Product/SharedProject/UIThread.cs

```diff
@@ -32,25 +32,19 @@ class UIThread : UIThreadBase, IDisposable {
         private readonly JoinableTaskFactory _factory;
         private readonly bool _needDispose;
 
-        private UIThread() {
-            try {
-                _factory = ThreadHelper.JoinableTaskFactory;
-                _context = _factory.Context;
+        public UIThread(JoinableTaskFactory joinableTaskFactory) {
+            if (joinableTaskFactory != null) {
+                _factory = joinableTaskFactory;
+                _context = joinableTaskFactory.Context;
                 Trace.TraceInformation("Using TID {0}:{1} as UI thread", _context.MainThread.ManagedThreadId, _context.MainThread.Name ?? "(null)");
-            } catch (NullReferenceException) {
+            } else {
                 _needDispose = true;
                 _context = new JoinableTaskContext();
                 _factory = new JoinableTaskFactory(_context);
```

Comment:
Do we not need to use the instance on `ThreadHelper` if it's already there? We need to share the context with the rest of VS, right?

Reply:
Yes, recommendation is to avoid `ThreadHelper`.

Reply:
So everyone just creates their own new joinable context? How does that work?

Reply:
Not context, but collection, factory, associated sync context and so on. If I recall correctly, recommendation for package-specific factories was there from the very beginning, but no one followed it because there was `ThreadHelper.JoinableTaskFactory`. So now they have implemented it by themselves. When package is disposed, it waits for all the tasks to complete:
```cs
        protected override void Dispose(bool disposing)
        {
            // Let any in-flight async set sites know they should just abort whatever they are doing.
            this.disposeCancellationTokenSource.Cancel();

            // Join on the tasks BEFORE we unregister/Dispose of the service instances, in case one of those services kicked off async work
            // that relies on said service. If we dispose services before waiting one of those tasks could encounter an ObjectDisposedException.
            try
            {
                ThreadHelper.JoinableTaskFactory.Run(this.JoinableTaskCollection.JoinTillEmptyAsync);
            }
```